### PR TITLE
scripts: respect environment variable AUR_COLOR

### PR DESCRIPTION
**Pull request by rafasc**
When color options are not specified, color ouput unless stdout or
stderr is not connected to a terminal.

With --color and --no-color, the colorization can be forced on or off.

### CODE REVIEW

**Comment by Morganamilo** [comment]
Wouldn't it make more sense to give some opt parsing to aur.in and make this a common option?

**Comment by rafasc** [comment]
One reason for using this approach is that scripts can be used independently of the main wrapper.

Teaching `aur.in` some option parsing for this would imply passing those variables along as exported variables, which to my knowledge, @AladW is not a fan of.

see https://github.com/AladW/aurutils/pull/448#issuecomment-417324408
> * I prefer using command-line options and stdin/stdout for IPC due to exported variables being [historically problematic](https://lists.archlinux.org/pipermail/aur-general/2015-August/031314.html) for AUR helpers.

**Comment by rafasc** [comment]
One small flaw I forgot to mention, is that `usage()` messages printed in the error handling part of `parseopts.sh` will never be colorized with this implementation, because the decision is done after option parsing.

That could either be handled separate with a subshell, or ignored, or even removed as parseopts.sh already prints the information about the unrecognized option.
```
if ! parseopts "$opt_short" "${opt_long[@]}" "${opt_hidden[@]}" -- "$@"; then
     ( [[ -t 2 ]] && colorize; usage )
fi```

**Comment by AladW** [comment]
I don't mind exported variables as long as they're optional.

On Tue, 02 Apr 2019 14:04:39 -0700
rafasc <notifications@github.com> wrote:

> One reason for using this approach is that scripts can be used independently of the main wrapper.
> 
> Teaching `aur.in` some option parsing for this would imply passing those variables along as exported variables, which to my knowledge, @AladW is not a fan of.
> 
> see https://github.com/AladW/aurutils/pull/448#issuecomment-417324408
> > * I prefer using command-line options and stdin/stdout for IPC due to exported variables being [historically problematic](https://lists.archlinux.org/pipermail/aur-general/2015-August/031314.html) for AUR helpers. 
> 
> 
> 
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/pull/557#issuecomment-479201732


-- 
Alad Wenter <alad@archlinux.org>

**Comment by rafasc** [comment]
Pushed an alternative version that makes use of environment variables instead.

The good: is that `$aur --color command --flags;` will enable color globally across the entire pipepline being called.
The bad and ugly: Without writing our own option parser, we need resort to hacks to make it work.

https://github.com/AladW/aurutils/blob/8f1cda7ca80235892a03b2cc18a20f19471b9045/aur.in#L32-L48

Side effect: Completions were written on the expectation that `aur(1)` doesn't have option handling, so they'll stop working if any `aur(1)` flag is used.

Suggestions welcome.

File: aur.in [hunk 1]
```diff
@@ -20,6 +21,46 @@ if [[ -z $1 ]]; then
...
+
+if ! parseopts "$opt_short" "${opt_long[@]}" "${opt_hidden[@]}" -- "${aur_args[@]}" -- "$@" ; then
+    usage
+fi
+set -- "${OPTRET[@]}"
+
+while true; do
+    case "$1" in
+        --color)        USE_COLOR=always ;;
+        --no-color)     USE_COLOR=never  ;;
```
Comment (AladW) [inline aur.in]: Shouldn't there be an `export` here?

**Comment by AladW** [comment]
I guess all those issues would be solved by simply not having a `--color` option, leaving everything to environment variables. These would only be set if the user wants to deviate from the default "auto" behavior.

File: aur.in (hunk 1 above)
Comment (rafasc) [reply aur.in]: I exported the variable at the top of the file.
https://github.com/AladW/aurutils/blob/8f1cda7ca80235892a03b2cc18a20f19471b9045/aur.in#L5

**Comment by rafasc** [comment]
I am ok with dropping the `--color` for now, but I think we will eventually come back to this limitation. Any options meant to be used across the entire pipeline, e.g. `--exec-path`, `--color`, `--repo`,  are potential candidates for `aur(1)` flags. 

But maybe we can cross that bridge if we get there.

**Comment by rafasc** [comment]
Dropped the commit that implemented the `aur` flags, documented the `USE_COLOR` on `aur.1`, rebased against master, and changed the title of the PR.

**Comment by AladW** [comment]
As a last change, perhaps namespace this to `AUR_COLOR` or `AUR_USE_COLOR`?

**Comment by rafasc** [comment]
Went with `AUR_COLOR`, and also removed intermediate lowercased variable.

**Comment by Earnestly** [comment]
For what it's worth, there is this effort: https://no-color.org/

**Comment by rafasc** [comment]
@Earnestly I am aware of that effort. It's a shame it is so focused on disabling color and not about controlling it. There are others ["standards"](https://xkcd.com/927/) that didn't get as much traction like https://bixense.com/clicolors/ for example. 
I would be more willing to comply if it was `$USE_COLOR` with values `[no,yes,auto] / [never,always,auto] / [0,1,2]`.

Anyways, in case we decide respect `$NO_COLOR`, using the following should be enough to be compliant.
```
AUR_COLOR=${NO_COLOR:+never}
AUR_COLOR=${AUR_COLOR:-auto}
```

File: lib/aur-build [hunk 2]
```diff
@@ -47,7 +48,7 @@ source /usr/share/makepkg/util/util.sh
 source /usr/share/makepkg/util/message.sh
 source /usr/share/makepkg/util/parseopts.sh
 
-if [[ -t 2 && ! -o xtrace ]]; then
+if [[ ( -t 1 && -t 2 && ! -o xtrace && $AUR_COLOR == auto ) || $AUR_COLOR == always ]] ; then
```
Comment (AladW) [inline lib/aur-build]: The manual update mentions `AUR_COLOR=never` or any unrecognized value. Should `never` be explicitely checked here, or the mention removed from `aur(1)`?

File: man1/aur.1 [hunk 3]
```diff
@@ -35,6 +35,18 @@ A different path may be set at build time through the
 macro, for example
 .IR /usr/local/lib/aurutils:/usr/lib/aurutils .
 
+Output colorization can be controlled by setting the environment variable
```
Comment (AladW) [inline man1/aur.1]: I would move this section to an `ENVIRONMENT` section, compare `man-pages(7)`

**Comment by AladW** [comment]
Something like `AUR_COLOR=${NO_COLOR:+never}` could also be used in the user's environment or shell file without requiring us to explicitely support it. Otherwise I agree having a binary "use color or don't" is not ideal.

File: lib/aur-build [hunk 4]
```diff
@@ -6,6 +6,7 @@ readonly argv0=build
 readonly startdir=$PWD
 readonly XDG_CONFIG_HOME=${XDG_CONFIG_HOME:-$HOME/.config}
 readonly PS4='+(${BASH_SOURCE}:${LINENO}): ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+AUR_COLOR=${AUR_COLOR:-auto}
```
[the remark this answers is not in the document]
Comment (rafasc) [reply lib/aur-build]: This was a left over from the approach where I was allowing `--color` and `--no-color` to modify the variable value. Will change.

File: man1/aur.1 (hunk 3 above)
Comment (rafasc) [reply man1/aur.1]: Will change.

File: lib/aur-build (hunk 2 above)
Comment (rafasc) [reply lib/aur-build]: The documentation is the API, the fact that we don't need to explicitly check for the 'never' value is just an implementation detail.

But I guess we can check for the 'never' value just for the sake of clarity.

``` bash
if [[ ( $AUR_COLOR == auto && -t 1 && -t 2 && ! -o xtrace ) || $AUR_COLOR == always || $AUR_COLOR != never ]] ; then
    colorize
fi
```
What do you think?

Comment (rafasc) [reply lib/aur-build]: `AUR_COLOR=${AUR_COLOR:-auto}` should be `AUR_COLOR=${AUR_COLOR-auto}`.

Comment (AladW) [reply lib/aur-build]: OK, let's keep the original approach then. (Though I'm not sure how `AUR_COLOR == always` is redundant, if you say "the last two tests".)

Comment (rafasc) [reply lib/aur-build]: Poor choice of words on my part. Only the last test is redundant. What I was trying to say is that `$AUR_COLOR == always && $AUR_COLOR != never`will only evaluate to true if the value is `always`, which is already being tested by the first part.

Comment (AladW) [reply lib/aur-build]: Yeah, if we accept any values for `AUR_COLOR` with only auto/always enabling color, such a check can only be redundant. (I guess the alternative is an interface where other values than auto/always/never cause an error, but eh.)

**Comment by AladW** [comment]
👍

**Pull request merged**